crates/oxc_query/src/edges.rs

```diff
@@ -0,0 +1,2169 @@
+        super::get_span(contexts)
+    }
+
+    pub(super) fn type_<'a, 'b: 'a>(
+        contexts: ContextIterator<'a, Vertex<'b>>,
+        resolve_info: &ResolveEdgeInfo,
+    ) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+        super::type_annotation::type_(contexts, resolve_info)
+    }
+}
+
+pub(super) fn resolve_type_edge<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+    edge_name: &str,
+    _parameters: &EdgeParameters,
+    resolve_info: &ResolveEdgeInfo,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+    match edge_name {
+        "span" => type_::span(contexts, resolve_info),
+        _ => {
+            unreachable!("attempted to resolve unexpected edge '{edge_name}' on type 'Type_'")
+        }
+    }
+}
+
+mod type_ {
+    use trustfall::provider::{
+        ContextIterator, ContextOutcomeIterator, ResolveEdgeInfo, VertexIterator,
+    };
+
+    use super::super::vertex::Vertex;
+
+    pub(super) fn span<'a, 'b: 'a>(
+        contexts: ContextIterator<'a, Vertex<'b>>,
+        _resolve_info: &ResolveEdgeInfo,
+    ) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+        super::get_span(contexts)
+    }
+}
+
+pub(super) fn resolve_url_edge<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+    edge_name: &str,
+    _parameters: &EdgeParameters,
+    resolve_info: &ResolveEdgeInfo,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+    match edge_name {
+        "search_parameter" => url::search_parameter(contexts, resolve_info),
+        _ => {
+            unreachable!("attempted to resolve unexpected edge '{edge_name}' on type 'URL'")
+        }
+    }
+}
+
+mod url {
+    use trustfall::provider::{
+        resolve_neighbors_with, ContextIterator, ContextOutcomeIterator, ResolveEdgeInfo,
+        VertexIterator,
+    };
+
+    use super::super::vertex::Vertex;
+    use crate::vertex::SearchParameterVertex;
+
+    pub(super) fn search_parameter<'a, 'b: 'a>(
+        contexts: ContextIterator<'a, Vertex<'b>>,
+        _resolve_info: &ResolveEdgeInfo,
+    ) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+        resolve_neighbors_with(contexts, |v| {
+            Box::new(
+                v.as_url()
+                    .unwrap_or_else(|| {
+                        panic!("expected to have a url vertex, instead have: {v:#?}")
+                    })
+                    .query_pairs()
+                    .map(|(key, value)| {
+                        Vertex::SearchParameter(
+                            SearchParameterVertex {
+                                key: key.to_string(),
+                                value: value.to_string(),
+                            }
+                            .into(),
+                        )
+                    })
+                    .collect::<Vec<_>>()
+                    .into_iter(),
```

Comment:
this collect then into_iter() is unavoidable because the (key,value) pairs have a weird lifetime that requires consuming them right then and there